File: Data Structures & Algorithms/combination-target-sum-ii/submission-3.java
class Solution {
    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        Arrays.sort(candidates);
        List<List<Integer>> res = new ArrayList<>();
        backtrack(candidates, 0, target, new ArrayList<>(), res);
        return res;
    }
    void backtrack(int[] nums, int i, int target, List<Integer> ds, List<List<Integer>> res){
    
        if(target == 0){
            res.add(new ArrayList<>(ds));
            return;
        }

        for(int j=i; j<nums.length; j++){
            if(j > i && nums[j-1] == nums[j]){
                continue;
            }
            if(target - nums[j] < 0){
                break;
            }
            ds.add(nums[j]);
            backtrack(nums, j+1, target-nums[j], ds, res);
            ds.remove(ds.size()-1);
        }
    }
}

File: Data Structures & Algorithms/partition-equal-subset-sum/submission-3.py
class Solution:
    def canPartition(self, nums: List[int]) -> bool:
        if sum(nums) % 2:
            return False
        target = sum(nums) // 2
        dp = [[-1] * (target+1) for _ in range(len(nums)+1)]

        def dfs(i, target):
            if target == 0:
                return True
            if i >= len(nums) or target < 0:
                return False
            if dp[i][target] != -1:
                return dp[i][target]

            dp[i][target] = dfs(i+1, target-nums[i]) or dfs(i+1, target)

            return dp[i][target]
        
        return dfs(0,target)